### PR TITLE
pd-client: reponse request when get stream error

### PR DESCRIPTION
**Pull request by nolouch**
ref: https://github.com/pingcap/tidb/pull/3393
we should response request when meeting stream error.
PTAL @disksing @siddontang @zimulala

### CODE REVIEW

File: pd-client/client.go [hunk 1]
```diff
@@ -297,6 +297,13 @@ func (c *client) tsLoop() {
 			if err != nil {
 				log.Errorf("[pd] create tso stream error: %v", err)
 				cancel()
+
+				n := len(c.tsoRequests)
+				for i := 0; i < n; i++ {
+					req := <-c.tsoRequests
+					req.done <- errors.Trace(err)
+				}
+
```
Comment (coocood) [inline pd-client/client.go]: Is it possible that new requests send to `tsoRequests` channel after responded all existing requests?

Comment (siddontang) [inline pd-client/client.go]: The new will receive another error after retrying connecting stream again.

Btw I see some places use the same code, can you extract these to a function?

Comment (nolouch) [inline pd-client/client.go]: yep, new requests will be received and try to get stream again.

**Comment by siddontang** [comment]
Any test to ensure it can work?

**Comment by tiancaiamao** [comment]
I'm worry about the reusing of tsoRequest
https://github.com/pingcap/pd/pull/656/files#diff-d7ad4791175fcda5c05f47ecfe3df530R330 may affect by this change.
The code is more and more complex and go beyond my comprehension.
@disksing @nolouch

**Comment by siddontang** [comment]
any update?

File: pd-client/client.go [hunk 2]
```diff
@@ -380,15 +381,19 @@ func (c *client) finishTSORequest(requests []*tsoRequest, physical, firstLogical
 	}
 }
 
-func (c *client) Close() {
-	c.cancel()
-	c.wg.Wait()
-
+func (c *client) finishExistTSORequest(err error) {
```
Comment (disksing) [inline pd-client/client.go]: How about rename to `revokeRequests()`?

File: pd-client/client.go [hunk 3]
```diff
@@ -380,15 +381,19 @@ func (c *client) finishTSORequest(requests []*tsoRequest, physical, firstLogical
...
 	n := len(c.tsoRequests)
 	for i := 0; i < n; i++ {
 		req := <-c.tsoRequests
 		req.done <- errors.Trace(errClosing)
```
Comment (disksing) [inline pd-client/client.go]: errClosing -> err

**Comment by nolouch** [comment]
@siddontang  I try to add the unit test. I test to close the server but grpc server not stop, I also get stream success. But I test it in local environment, it works.

**Comment by coocood** [comment]
LGTM

**Pull request merged**